File: pie/src/submit.rs
//! Submit command implementation for the Pie CLI.
//!
//! This module implements the `pie submit` subcommand for submitting inferlets
//! to an existing running Pie engine instance.

use crate::{engine, path};
use anyhow::Context;
use anyhow::Result;
use clap::Args;
use libpie::auth;
use std::{fs, path::PathBuf};

/// Arguments for the `pie submit` command.
#[derive(Args, Debug)]
pub struct SubmitArgs {
    /// Path to the .wasm inferlet file.
    #[arg(value_parser = path::expand_tilde)]
    pub inferlet: PathBuf,
    /// Path to a custom TOML configuration file.
    #[arg(long)]
    pub config: Option<PathBuf>,
    /// The network host to connect to.
    #[arg(long)]
    pub host: Option<String>,
    /// The network port to connect to.
    #[arg(long)]
    pub port: Option<u16>,
    /// Authentication secret for connecting to the server.
    #[arg(long)]
    pub auth_secret: Option<String>,
    /// Arguments to pass to the inferlet after `--`.
    #[arg(last = true)]
    pub arguments: Vec<String>,
}

/// Handles the `pie submit` command.
///
/// This function:
/// 1. Reads configuration from the specified config file or default config
/// 2. Creates a client configuration from config and command-line arguments
/// 3. Connects to the existing Pie engine server
/// 4. Submits the specified inferlet with the provided arguments
/// 5. Waits for the inferlet to finish execution and prints the result
pub async fn handle_submit_command(
    config_path: Option<PathBuf>,
    host: Option<String>,
    port: Option<u16>,
    auth_secret: Option<String>,
    inferlet_path: PathBuf,
    arguments: Vec<String>,
) -> Result<()> {
    // Read config file only if when any parameter is missing
    let config_file = if host.is_none() || port.is_none() || auth_secret.is_none() {
        let config_str = match config_path {
            Some(path) => fs::read_to_string(&path)
                .with_context(|| format!("Failed to read config file at {:?}", path))?,
            None => fs::read_to_string(&crate::path::get_default_config_path()?).context(
                "Failed to read default config file. Try running `pie config init` first.",
            )?,
        };
        Some(toml::from_str::<crate::config::ConfigFile>(&config_str)?)
    } else {
        None
    };

    // Prefer command-line arguments and use config file values if not provided
    let host = host
        .or_else(|| config_file.as_ref().and_then(|cfg| cfg.host.clone()))
        .unwrap_or_else(|| "127.0.0.1".to_string());

    let port = port
        .or_else(|| config_file.as_ref().and_then(|cfg| cfg.port))
        .unwrap_or(8080);

    let auth_secret = auth_secret
        .or_else(|| config_file.as_ref().and_then(|cfg| cfg.auth_secret.clone()))
        .unwrap_or_else(engine::generate_random_auth_secret);

    // Initialize the JWT secret for authentication
    auth::init_secret(&auth_secret);

    // Create client configuration
    let client_config = engine::ClientConfig { host, port };

    // Submit the inferlet to the existing server
    engine::submit_inferlet_and_wait(&client_config, inferlet_path, arguments).await?;

    Ok(())
}

File: client/python/src/pie_client_cli/submit.py
"""Submit command implementation for the Pie CLI.

This module implements the `pie-cli submit` subcommand for submitting inferlets
to an existing running Pie engine instance.
"""

import shutil
import subprocess
import tempfile
import tomllib
from pathlib import Path
from typing import Optional

import typer

from . import engine


def parse_manifest(manifest_content: str) -> tuple[str, str, str]:
    """Parse the manifest to extract namespace, name, and version.

    Args:
        manifest_content: The TOML manifest content as a string.

    Returns:
        A tuple of (namespace, name, version).

    Raises:
        ValueError: If the manifest is missing required fields or has invalid format.
    """
    manifest = tomllib.loads(manifest_content)

    package = manifest.get("package")
    if package is None:
        raise ValueError("Manifest missing [package] section")

    full_name = package.get("name")
    if full_name is None:
        raise ValueError("Manifest missing package.name field")

    version = package.get("version")
    if version is None:
        raise ValueError("Manifest missing package.version field")

    # Parse "namespace/name" format
    parts = full_name.split("/", 1)
    if len(parts) != 2:
        raise ValueError(
            f"Invalid package.name format '{full_name}': expected 'namespace/name'"
        )

    return parts[0], parts[1], version


def compose_components(
    program_path: Path, library_paths: list[Path], output_path: Path
) -> None:
    """Compose a program with multiple libraries using wac CLI.

    Uses `wac plug` command to link libraries into the main program.
    Libraries are linked sequentially in the order provided.

    Args:
        program_path: Path to the main program WASM file.
        library_paths: List of paths to library WASM files.
        output_path: Path where the composed WASM binary will be written.

    Raises:
        RuntimeError: If wac CLI is not available or composition fails.
        FileNotFoundError: If program or library files are not found.
    """
    # Check if wac is available
    try:
        result = subprocess.run(
            ["wac", "--version"],
            capture_output=True,
            text=True,
        )
        if result.returncode != 0:
            raise RuntimeError("wac CLI is not available")
    except FileNotFoundError:
        raise RuntimeError(
            "wac CLI is not installed. Install it with: cargo install wac-cli"
        )

    if not program_path.exists():
        raise FileNotFoundError(f"Program file not found: {program_path}")

    # Start with the main program
    current_wasm = program_path

    for library_path in library_paths:
        if not library_path.exists():
            raise FileNotFoundError(f"Library file not found: {library_path}")

        # Compose using wac plug
        # wac plug --plug <library.wasm> <main.wasm> -o <output.wasm>
        with tempfile.NamedTemporaryFile(suffix=".wasm", delete=False) as tmp:
            tmp_output = Path(tmp.name)

        try:
            result = subprocess.run(
                [
                    "wac",
                    "plug",
                    "--plug",
                    str(library_path),
                    str(current_wasm),
                    "-o",
                    str(tmp_output),
                ],
                capture_output=True,
                text=True,
            )

            if result.returncode != 0:
                raise RuntimeError(
                    f"wac plug failed for {library_path}:\n{result.stderr}"
                )

            # Clean up previous temp file if it's not the original program
            if current_wasm != program_path:
                current_wasm.unlink()

            current_wasm = tmp_output
        except Exception:
            tmp_output.unlink(missing_ok=True)
            raise

    # Move the final composed binary to the output path
    if current_wasm != program_path:
        current_wasm.rename(output_path)
    else:
        # No libraries were linked, just copy the original
        shutil.copy(program_path, output_path)


def handle_submit_command(
    inferlet: Optional[str] = None,
    path: Optional[Path] = None,
    manifest: Optional[Path] = None,
    config: Optional[Path] = None,
    host: Optional[str] = None,
    port: Optional[int] = None,
    username: Optional[str] = None,
    private_key_path: Optional[Path] = None,
    detached: bool = False,
    link: Optional[list[Path]] = None,
    arguments: Optional[list[str]] = None,
) -> None:
    """Handle the `pie-cli submit` command.

    You can specify an inferlet either by registry name or by path (mutually exclusive):

    - By registry: pie-client submit std/text-completion@0.1.0
    - By path: pie-client submit --path ./my_inferlet.wasm --manifest ./Pie.toml

    Steps:
    1. Creates a client configuration from config file and command-line arguments
    2. Connects to the Pie engine server
    3. If using path and libraries are specified, composes them with the inferlet using wac
    4. Uploads the composed inferlet if not already on server (path mode only)
    5. Launches the inferlet with the provided arguments
    6. In non-detached mode, streams the inferlet output with signal handling
    """
    # Validate at least one of inferlet or path is provided
    if inferlet is None and path is None:
        typer.echo("Error: Specify an inferlet name or --path", err=True)
        raise typer.Exit(1)

    # Handle the case where --path is used with -- separator
    # Positional args after -- get captured as `inferlet` first, so
    # prepend it to `arguments` instead
    if inferlet is not None and path is not None:
        arguments = [inferlet] + (arguments or [])
        inferlet = None

    # Validate manifest is provided when using --path
    if path is not None and manifest is None:
        typer.echo("Error: --manifest is required when using --path", err=True)
        raise typer.Exit(1)

    link = link or []
    arguments = arguments or []

    client_config = engine.ClientConfig.create(
        config_path=config,
        host=host,
        port=port,
        username=username,
        private_key_path=private_key_path,
    )

    client = engine.connect_and_authenticate(client_config)

    try:
        if path is not None:
            # Launch from local file path
            if not path.exists():
                raise FileNotFoundError(f"Inferlet file not found: {path}")

            if not manifest.exists():
                raise FileNotFoundError(f"Manifest file not found: {manifest}")

            manifest_content = manifest.read_text()

            # Parse the manifest to extract namespace, name, and version
            namespace, name, version = parse_manifest(manifest_content)
            inferlet_name = f"{namespace}/{name}@{version}"

            typer.echo(f"Inferlet: {inferlet_name}")

            # If libraries are specified, compose them and always upload
            if link:
                with tempfile.NamedTemporaryFile(suffix=".wasm", delete=False) as tmp:
                    composed_path = Path(tmp.name)
                    try:
                        compose_components(path, link, composed_path)
                        engine.upload_program(client, composed_path, manifest)
                    finally:
                        composed_path.unlink(missing_ok=True)
                typer.echo("✅ Inferlet upload successful.")
            # No composition - check if program already exists before uploading
            else:
                if not engine.program_exists(client, inferlet_name, path, manifest):
                    engine.upload_program(client, path, manifest)
                    typer.echo("✅ Inferlet upload successful.")
                else:
                    typer.echo("Inferlet already exists on server.")

            # Launch the instance
            instance = engine.launch_instance(
                client,
                inferlet_name,
                arguments,
                detached,
            )
        else:
            # Launch from registry
            if link:
                typer.echo(
                    "Warning: --link option is ignored when launching from registry",
                    err=True,
                )

            typer.echo(f"Launching from registry: {inferlet}")

            instance = engine.launch_instance_from_registry(
                client,
                inferlet,
                arguments,
                detached,
            )

        typer.echo(f"✅ Inferlet launched with ID: {instance.instance_id}")

        if not detached:
            engine.stream_inferlet_output(instance, client)

    finally:
        engine.close_client(client)
